### ISSUE TITLE
"html_fragment" and "envir" functionality from Rmd?

### ISSUE THREAD

**Issue by richarddmorey** [quarto-dev/quarto-r#73]
I often used two features of Rmarkdown that I can't find equivalents for:
* the ability to create html fragments (html documents without headers) (with https://rmarkdown.rstudio.com/html_fragment_format.html), and
* the ability to compile a document with `rmarkdown::render` in a specified environment (via the `envir` argument) so that I could pass data objects to the document.

I can't find equivalents to these in quarto. I thought `standalone: false` might do it, based on the description ("Produce output with an appropriate header and footer (e.g. a standalone HTML, LaTeX, TEI, or RTF file, not a fragment)", from https://quarto.org/docs/reference/formats/html.html) but as far as I can tell, that option does not change the compilation of my documents at all.

Are there equivalents to these? Thanks!

**Comment by cderv** [comment]
> the ability to create html fragments

Regarding fragments, see discussion in https://github.com/quarto-dev/quarto-cli/discussions/3529. 
There is not yet a specific format like `html_fragment()` but it would be quite easy to create one by just providing a template with only a body (See [template argument for Quarto HTML format](https://quarto.org/docs/reference/formats/html.html#rendering) and see [the one in **rmarkdown**](https://github.com/rstudio/rmarkdown/blob/main/inst/rmd/fragment/default.html).

> the ability to compile a document with rmarkdown::render in a specified environment (via the envir argument) so that I could pass data objects to the document.

Regarding `envir`, see some more detailed answer in https://github.com/quarto-dev/quarto-r/issues/63. In short, Quarto will call R to execute the R cells but that is all. For the rest, it is not in R. `quarto render` can't take a `envir` argument and `quarto::quarto_render()` is a wrapper aroung the CLI `quarto`. 

To pass some data to your R document, you need to use some files and load data in first chunks of your document. Every R code is run in a clean new session with Quarto - unless R Markdown where you could do stuff before calling `rmarkdown::render()` or even pass a specific `envir` as advanced usage. 

> I thought standalone: false might do it, based on the description ("Produce output with an appropriate header and footer (e.g. a standalone HTML, LaTeX, TEI, or RTF file, not a fragment)", from [quarto.org/docs/reference/formats/html.html](https://quarto.org/docs/reference/formats/html.html)) but as far as I can tell, that option does not change the compilation of my documents at all.

You can produce minimal HTML using the `minimal: true` config (https://quarto.org/docs/output-formats/html-basics.html#minimal-html) but you will loose some of the Quarto features. Same if you don't have header. 

If you only want to convert from md to HTML without any Quarto feature, you can use pandoc, even from Quarto command 
Then you can use the embeded Pandoc to your conversion
````
quarto pandoc ...
````

Just curious - why do you need equivalent of HTML fragment in Quarto ? Just to run R cells and get the body part of the HTML ?

**Comment by richarddmorey** [comment]
> Just curious - why do you need equivalent of HTML fragment in Quarto ? Just to run R cells and get the body part of the HTML ?

I want to inject the result directly into another containing webpage (eg, as you would with a blog post), so I don't want the header. rmarkdown::render is much faster with html_fragment (and I'll be rendering these on the fly).  The containing webpage already contains the necessary headers.

**Comment by cderv** [comment]
Ok thanks. So you don't want any specific Quarto HTML feature then ? Just to convert Markdown to HTML with code cell results inserted. 

Currently this is not built in and would need to be implemented using an extension format probably using a custom template.

**Comment by richarddmorey** [comment]
> So you don't want any specific Quarto HTML feature then ? Just to convert Markdown to HTML with code cell results inserted.

Yes, that's right. The two benefits are 1) clean HTML that doesn't conflict (code-wise or style-wise) with the parent document, and 2) speed of compilation. You could put the Quarto-compiled complete document in an iframe, but then you'd lose the style coherence with the parent document.

I tried this with a minimal HTML template containing only `$body$` and `minimal: true` but I still got lots of extraneous content (this is the default Quarto document in RStudio):

```
<html><head>
<link href="Untitled_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="Untitled_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"></head><body><h2 id="quarto">Quarto</h2>
<p>Quarto enables you to weave together content and executable code into a finished document. To learn more about Quarto see <a href="https://quarto.org" class="uri">https://quarto.org</a>.</p>
<h2 id="running-code">Running Code</h2>
<p>When you click the <strong>Render</strong> button a document will be generated that includes both content and the output of embedded code. You can embed code like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="dv">1</span> <span class="sc">+</span> <span class="dv">1</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2</code></pre>
</div>
</div>
<p>You can add options to executable code like this</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4</code></pre>
</div>
</div>
<p>The <code>echo: false</code> option disables the printing of code (only output is displayed).</p>
</body></html>
```
I'm somewhat surprised that `$body$` in a template gives me more than the `<body>`.

If I render the equivalent Rmd document as an `html_fragment`, I get:

```
<div id="quarto" class="section level2">
<h2>Quarto</h2>
<p>Quarto enables you to weave together content and executable code into
a finished document. To learn more about Quarto see <a href="https://quarto.org" class="uri">https://quarto.org</a>.</p>
</div>
<div id="running-code" class="section level2">
<h2>Running Code</h2>
<p>When you click the <strong>Render</strong> button a document will be
generated that includes both content and the output of embedded code.
You can embed code like this:</p>
<pre class="r"><code>1 + 1</code></pre>
<pre><code>## [1] 2</code></pre>
<p>You can add options to executable code like this</p>
<pre><code>## [1] 4</code></pre>
<p>The <code>echo: false</code> option disables the printing of code
(only output is displayed).</p>
</div>
```

(I also note that the quarto ducument takes over 2s to compile, whereas the Rmd `html_fragment` takes less than .2s on my system.)

**Comment by dragonstyle** [comment]
FWIW I agree that it is surprising to see the link elements present. We are clearly inadvertently including those when they should be excused (and this is a bug).

**Comment by cderv** [comment]
Thanks @richarddmorey ! 

@dragonstyle I agree also that some insertion seems they should not be there - However, I think we've made `minimal` and `theme` disconnected. 

@richarddmorey can you also try with `theme: none` (https://quarto.org/docs/output-formats/html-themes.html) This should remove the `<head>` part that is not needed here.

I believe both options `minimal` and `theme` are needed to have Quarto not add anything required for its feature. 

I think we could make a format extensions for this purpose maybe, to mimic a `html_fragment` in R Markdown. @dragonstyle what do you think ? I'll probably give it a try. 

> (I also note that the quarto document takes over 2s to compile, whereas the Rmd html_fragment takes less than .2s on my system.)

Regarding speed, Quarto will initialize a lot at each rendering - some which are definitely not needed in your case. We'll keep working on speed improvement in near future. You could try 1.3 pre release if not already, it is possible that this has improve.

**Comment by richarddmorey** [comment]
> can you also try with theme: none (https://quarto.org/docs/output-formats/html-themes.html) This should remove the <head> part that is not needed here.

@cderv - adding `theme: none` does not seem to have any effect (I get the same document as without it; quarto 1.2, R 4.2.1). Happy to share more details and a reproducible example if necessary.

**Comment by cderv** [comment]
Thanks for trying. 

Could you try `highlight-style: none` now ? I believe this should remove `quarto-syntax-highlighting.css`

Just trying to help with workaround for now but it seems the first CSS cannot be prevented for now. 
Hopefully #3933 will fix this. I believe we should not include anything when not `<head>` and `$header-includes$ variable ` is in the template - or a least provide a way for making clean fragment HTML .

**Comment by jimjam-slam** [comment]
I only recently learned about the `keep-md` option in Quarto. That could help here!

1. render to HTML with `keep-md` enabled, which produces both a full HTML file and a Markdown file;
2. then use Quarto's bundled version of Pandoc to overwrite the HTML with an HTML fragment from the intermediary Markdown file.

`test.qmd`:

````
---
title: Hey there
author: James
keep-md: true
---

Oh hi! 2 + 2 is:

```{r}
2 + 2
```

Isn't that right?
````

Command:

```sh
quarto render test.qmd && quarto pandoc test.md -o test.html
```

Output `test.html`:

```html
<p>Oh hi! 2 + 2 is:</p>
<div class="cell">
<div class="sourceCode" id="cb1"><pre
class="sourceCode r cell-code"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="dv">2</span> <span class="sc">+</span> <span class="dv">2</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4</code></pre>
</div>
</div>
<p>Isn’t that right?</p>
```

Is that what you're looking for?

**Comment by cderv** [comment]
@jimjam-slam this could indeed be a hackish workaround for current situation, by bypassing quarto convertion to HTML and relying only on Pandoc. Some features will be missing though and depending on the content, there could be some unresolved Quarto markdown specificity that happens during the Pandoc conversion to HTML. 

So maybe working for simple content but probably not "safe" for more complex document I think.

**Comment by philipshirk** [comment]
> Regarding fragments, see discussion in [quarto-dev/quarto-cli#3529](https://github.com/quarto-dev/quarto-cli/discussions/3529).
> There is not yet a specific format like `html_fragment()` but it would be quite easy to create one by just providing a template with only a body (See [template argument for Quarto HTML format](https://quarto.org/docs/reference/formats/html.html#rendering) and see [the one in **rmarkdown**](https://github.com/rstudio/rmarkdown/blob/main/inst/rmd/fragment/default.html).

I used the [rmd default.html file](https://github.com/rstudio/rmarkdown/blob/main/inst/rmd/fragment/default.html) as a Quarto template, but no html widgets (plotly plots or DT tables) are in the output.

**Comment by cderv** [comment]
> I used the [rmd default.html file](https://github.com/rstudio/rmarkdown/blob/main/inst/rmd/fragment/default.html?rgh-link-date=2024-01-16T19%3A49%3A07Z) as a Quarto template, but no HTML widgets (plotly plots or DT tables) are in the output.

I don't think this is Quarto-related. You won't get them also using `rmarkdown::html_fragment()` with `rmarkdown::render()` in a `.Rmd` file. HTML widgets require some external resources (external JS lib), and the `html_fragment()` format does not have `<head>` to put these into. It does not even have a `<body>` part. 

This is not made for that - Usage of ``html_fragment` is specific. So if the idea to use a similar template with Quarto is really good, I believe there is a question of usage and why you are looking to use this `default.html` in the first place if you expect some complex content in it.

**Comment by philipshirk** [comment]
Thank you for the explanation. My goal is to create a dashboard that shows the results of a single analysis, but to let the user select the dataset on which the analysis was performed. Because the analysis is time-consuming, I don't want to just create a shiny dashboard. I thought a solution would be to create an html fragment with the results of each analysis, then create a shiny dashboard that lets the user select which pre-compiled html fragment to display. I would prefer an html document rather than a shiny dashboard just so no backend server is required, but I haven't come up with a better option than to use a shiny dashboard that displays full html files in an iframe. (Although I now see that it is possible to create shiny cache that persists across sessions, so maybe that's the best option.)

**Comment by cderv** [comment]
Thanks @philipshirk - let's move this use case to a discussion item to continue. I'll open and ping you.
